Automattic/newspack-app-shell
# Main functionality setup

This introduces the main feature of the plugin.

## How it works

1. loads a JS script on the front-end, which:
1. hijacks navigation events
1. prevents native navigation where applicable (when a request is to another location on the site)
1. and instead fetches the content via XHR
1. replacing it, while keeping the persistent element (as well as any unchanged elements) intact

## How to test 

1. prepare the site/theme: 
    1. use `add/handle-app-shell` branch of the theme repo, then `npm run build` the theme assets
    1. in AMP plugin settings, switch to use `Transitional` mode
1. clone the repo to the plugins directory of your WP installation
1. install the dependencies: `composer install && npm i && npm run build`
1. activate the `Newspack App Shell` plugin
1. you should see an `Add Persistent element` item in the admin menu, click it 
1. switch to code editor and paste the following there:

```
<iframe width="100%" height="140px" scrolling="no" frameborder="no" allow="autoplay" src="https://w.soundcloud.com/player/?url=https%3A//api.soundcloud.com/tracks/155469656&amp;color=%23ff5500&amp;auto_play=false&amp;hide_related=false&amp;show_comments=true&amp;show_user=true&amp;show_reposts=false&amp;show_teaser=true&amp;visual=true"></iframe>
<style media="screen">
iframe { position: fixed; z-index: 999; bottom: 0; left: 0; }
</style>
```
7. publish the post
8. visit the website
9. observe a SoundCloud player glued to the bottom of the viewport. Click play
10. navigate the website:
    1. the music should not stop and the player should stay where is was
    1. but the website content should change as expected, but with nice animations
11. When you visit the admin panel again (or refresh), the new menu item should say "Edit Persistent element" and point to the just-created post

### That's how it should look like:

![Screen Capture on 2020-01-29 at 10-42-12](https://user-images.githubusercontent.com/7383192/73345275-1bef3800-4284-11ea-9ef4-a449e38745e4.gif)
